src/mame/sony_news/news_r4k.cpp

```diff
@@ -0,0 +1,1261 @@
+		ENTRY_PAD_SHIFT = 20,
+		ENTRY_PFNUM = 0xFFFFF
+	};
+
+	// APbus DMA page table entry structure
+	// See https://github.com/NetBSD/src/blob/faa527952534272f8f4737347d9fc8c38e070c9c/sys/arch/newsmips/apbus/dmac3reg.h#L71
+	struct apbus_pte
+	{
+		// Unused
+		uint32_t pad;
+		// Entry is OK to use
+		bool valid;
+		// DMA coherence enabled (don't care for emulation??)
+		bool coherent;
+		// Unused
+		uint32_t pad2;
+		// Page number (upper 20 bits of physical address)
+		uint32_t pfnum;
+	};
+
+	// Interrupts and other platform state
+	bool m_ram_map_shift = false;
+	bool m_int_state[6] = {false, false, false, false, false, false};
+	uint32_t m_inten[6] = {0, 0, 0, 0, 0, 0};
+	uint32_t m_intst[6] = {0, 0, 0, 0, 0, 0};
+	uint32_t escc1_int_status = 0;
+	memory_access<64, 3, 0, ENDIANNESS_BIG>::cache m_main_cache;
+
+	// Freerun timer (1us period)
+	// NetBSD source code corroborates the period (https://github.com/NetBSD/src/blob/229cf3aa2cda57ba5f0c244a75ae83090e59c716/sys/arch/newsmips/newsmips/news5000.c#L259)
+	emu_timer *m_freerun_timer;
+	uint32_t m_freerun_timer_val;
+
+	// Hardware timer TIMER0 (10ms period)
+	// TODO: Is this programmable somehow? 100Hz is what the NetBSD kernel expects.
+	emu_timer *m_timer0_timer;
+
+	// MIPS R4400 CPU
+	required_device<r4400_device> m_cpu;
+
+	// Memory
+	required_device<ram_device> m_ram;
+	required_device<ram_device> m_dma_ram;
+	required_device<ram_device> m_net_ram;
+
+	// ST Micro M48T02 Timekeeper NVRAM + RTC
+	required_device<m48t02_device> m_rtc;
+
+	// Zilog ESCC and Sony CXD8421Q ESCC1 serial controller
+	required_device<z80scc_device> m_escc;
+	required_device_array<rs232_port_device, 2> m_serial;
+
+	// Sony CXD8442Q APbus FIFOs
+	required_device<cxd8442q_device> m_fifo0;
+	required_device<cxd8442q_device> m_fifo1;
+
+	// Sony CXD8452AQ WSC-SONIC3 APbus interface
+	required_device<cxd8452aq_device> m_sonic3;
+
+	// SONIC ethernet controller
+	required_device<dp83932c_device> m_sonic;
+
+	// National Semiconductor PC8477B floppy controller
+	required_device<pc8477b_device> m_fdc;
+
+	// NEWS keyboard and mouse
+	required_device<news_hid_hle_device> m_hid;
+
+	// DMAC3 DMA controller
+	required_device<dmac3_device> m_dmac;
+
+	// HP SPIFI3 SCSI controllers
+	required_device<spifi3_device> m_scsi0;
+	required_device<spifi3_device> m_scsi1;
+	required_device<nscsi_bus_device> m_scsibus0;
+	required_device<nscsi_bus_device> m_scsibus1;
+
+	// Front panel
+	required_ioport m_dip_switch;
+	output_finder<6> m_led;
+
+	// Constants
+	static constexpr uint32_t ICACHE_SIZE = 0x4000;
+	static constexpr uint32_t DCACHE_SIZE = 0x4000;
+	static constexpr uint32_t SCACHE_SIZE = 0x100000;
+	static constexpr int FREERUN_FREQUENCY = 1000000; // Hz
+	static constexpr int TIMER0_FREQUENCY = 100;      // Hz
+	static constexpr uint32_t APBUS_DMA_MAP_ADDRESS = 0x14c20000;
+	static constexpr uint32_t APBUS_DMA_MAP_RAM_SIZE = 0x20000; // 128 kibibytes
+	const char *MAIN_MEMORY_DEFAULT = "64M";
```

Comment:
Should this be static as well?  Should it have another `const` to make the pointer `const` as well as the thing it points to?

Reply:
Done. I made it `constexpr` to avoid having to set it separately, let me know if you'd prefer it the other way (same applies to the others listed below).